rir/src/interpreter/cache.h

```diff
@@ -0,0 +1,126 @@
+#ifndef RIR_INTERPRETER_CACHE_H
+#define RIR_INTERPRETER_CACHE_H
+
+#include "instance.h"
+
+namespace rir {
+
+// must be a power of 2 for modulus using & CACHE_MASK to work
+#define MAX_CACHE_SIZE 64
+#define CACHE_MASK (MAX_CACHE_SIZE - 1)
+#define ACTIVE_BINDING_MASK (1 << 15)
+#define BINDING_LOCK_MASK (1 << 14)
+#define IS_ACTIVE_BINDING(b) ((b)->sxpinfo.gp & ACTIVE_BINDING_MASK)
+#define BINDING_IS_LOCKED(b) ((b)->sxpinfo.gp & BINDING_LOCK_MASK)
+
+static RIR_INLINE Immediate cacheIndex(Immediate index, bool smallCache) {
+    Immediate cidx;
+    if (smallCache)
+        cidx = index;
+    else
+        cidx = index & CACHE_MASK;
+    return cidx;
+}
+
+#ifdef CACHE_ON_R_STACK
+// TODO: Create a version with a cache on the R_stack instead of the C stack
+#else
+typedef struct {
+    SEXP loc;
+    Immediate idx;
```

Comment:
drive-by comment: this should not be needed anymore. if we have one fix location per <env,name> pair, then there is just no need to check if it is the correct name